### ISSUE TITLE
Spring boot loader cannot find classnames with accentuated char (french, UTF-8) in jarfiles

### ISSUE THREAD

**Issue by mercibe** [spring-projects/spring-boot#5194]
Hi,

My  Spring boot application depends on a library with Entities and Repositories (Spring data JPA) which contain Entities classname with accentuated french characters (UTF-8),  My current Spring Web application (4.2.4.RELEASE) works perfectly well with this library, but when I want to use it from a new Spring boot application (tested with 1.3.2 and 1.4.0-BUILD-SNAPSHOT, open JDK 8 on Linux) I get the following stacktrace (from the sample app attached to this report):

```
Caused by: java.io.FileNotFoundException: JAR entry debug/mercibe/entités/MotCl�.class not found in /home/benoitm/src/exploration/spring-boot-app/target/spring-boot-app-0.0.1-SNAPSHOT.jar!/lib/jar-utf8-classname-1.0-SNAPSHOT.jar
        at org.springframework.boot.loader.jar.JarURLConnection.throwFileNotFound(JarURLConnection.java:119)
        at 
org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:107)
        at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:175)
        at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:168)
        at org.springframework.core.type.classreading.SimpleMetadataReader.<init>(SimpleMetadataReader.java:50)
        at org.springframework.core.type.classreading.SimpleMetadataReaderFactory.getMetadataReader(SimpleMetadataReaderFactory.java:98)
        at org.springframework.core.type.classreading.CachingMetadataReaderFactory.getMetadataReader(CachingMetadataReaderFactory.java:102)
```

You will find attached a ZIP file containing 2 projects : 

  - jar-utf8-classname : a maven jar project with one entity and one repository with special chars.
  - spring-boot-app : an « empty » Spring boot app that depends on jar-utf8-classname

To reproduce : 

  - unzip [spring-boot-utf8-classname.zip](https://github.com/spring-projects/spring-boot/files/141388/spring-boot-utf8-classname.zip)
  - install jar-utf8-classname (cd jar-utf8-classname) : `mvn clean install`
  - run sample app (cd spring-boot-app) : `mvn clean package && java -jar target/spring-boot-app-0.0.1-SNAPSHOT.jar `

For info, « mot-clé » means « keyword » in French ;-)

I suppose that a reader somewhere should use UTF-8 to decode an inputstream, but where...  I noticed that it is the reference of MotClé in the repository interface declaration (`public interface MotCléRepository extends CrudRepository<MotClé, Long> {...}`) that triggered the error.

Thank you very much for your support.

Benoit

**Comment by wilkinsona** [comment]
Thank you for the detailed problem report and sample.

It's interesting that `entités` is handled correctly but that `MotClé` is not. This happens within `PathMatchingResourcePatternResolver` which is looking for entries matching `classpath*:debug/mercibe/entités/**/*.class`. The URL for the `debug.mercibe.entités` package is:

```
jar:file:/Users/awilkinson/Downloads/spring-boot-utf8-classname/spring-boot-app/target/spring-boot-app-0.0.1-SNAPSHOT.jar!/lib/jar-utf8-classname-1.0-SNAPSHOT.jar!/debug/mercibe/entit%c3%a9s/]
```
Note the encoding of the `é` into `%c3%a9`. It then finds the `MotClé` class in this package and creates a URL for it:

```
jar:file:/Users/awilkinson/Downloads/spring-boot-utf8-classname/spring-boot-app/target/spring-boot-app-0.0.1-SNAPSHOT.jar!/lib/jar-utf8-classname-1.0-SNAPSHOT.jar!/debug/mercibe/entit%c3%a9s/MotClé.class
```

Note the lack of encoding in `MotClé.class`. When an `InputStream` is opened for this URL, Boot's `JarURLConnection` fails to find the entry as it expects the `é` to have been encoded.

**Comment by wilkinsona** [comment]
Here's a possible fix. It updates `JarURLConnection.JarEntryName` to cope with characters in the spec that should have been encoded but that have not:

```diff
diff --git a/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarURLConnection.java b/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarURLConnection.java
index d2c9a62..d1e0f62 100644
--- a/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarURLConnection.java
+++ b/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarURLConnection.java
@@ -20,9 +20,11 @@ import java.io.ByteArrayOutputStream;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
+import java.net.URLEncoder;
 import java.net.URLStreamHandler;
 import java.util.jar.Manifest;

@@ -227,20 +229,38 @@ class JarURLConnection extends java.net.JarURLConnection {
                                return new AsciiBytes(source);
                        }
                        ByteArrayOutputStream bos = new ByteArrayOutputStream(length);
+                       write(source, bos);
+                       // AsciiBytes is what is used to store the JarEntries so make it symmetric
+                       return new AsciiBytes(bos.toByteArray());
+               }
+
+               private void write(String source, ByteArrayOutputStream out) {
+                       int length = source.length();
                        for (int i = 0; i < length; i++) {
                                int ch = source.charAt(i);
-                               if (ch == '%') {
-                                       if ((i + 2) >= length) {
-                                               throw new IllegalArgumentException("Invalid encoded sequence \""
-                                                               + source.substring(i) + "\"");
+                               if (ch > 127) {
+                                       try {
+                                               String encoded = URLEncoder.encode(String.valueOf((char) ch),
+                                                               "UTF-8");
+                                               write(encoded, out);
+                                       }
+                                       catch (UnsupportedEncodingException ex) {
+                                               throw new IllegalStateException(ex);
                                        }
-                                       ch = decodeEscapeSequence(source, i);
-                                       i += 2;
                                }
-                               bos.write(ch);
+                               else {
+                                       if (ch == '%') {
+                                               if ((i + 2) >= length) {
+                                                       throw new IllegalArgumentException(
+                                                                       "Invalid encoded sequence \"" + source.substring(i)
+                                                                                       + "\"");
+                                               }
+                                               ch = decodeEscapeSequence(source, i);
+                                               i += 2;
+                                       }
+                                       out.write(ch);
+                               }
                        }
-                       // AsciiBytes is what is used to store the JarEntries so make it symmetric
-                       return new AsciiBytes(bos.toByteArray());
                }

                private char decodeEscapeSequence(String source, int i) {
```

**Comment by mercibe** [comment]
Thank you very much for your fast reply.  I will refactor my classes right now since I am in a hurry but I will bring their original names back once the Spring Boot team will have applied the fix in a forthcoming release.

**Comment by mercibe** [comment]
For your information : I applied the fix wilkinsona provided on 1.3.2.RELEASE, built and deployed it on our local Maven proxy and it worked perfectly well with all our projects.

For the 1.3.2.RELEASE I just had to add a toString call (return of decode() function)

`return new AsciiBytes(bos.toByteArray()).toString();`

(JarURLConnection.java for 1.3.2.RELEASE  attached to this comment)

In my opinion it would deserve a 1.3.**3**.RELEASE ;-)
[JarURLConnection.zip](https://github.com/spring-projects/spring-boot/files/144817/JarURLConnection.zip)

**Comment by wilkinsona** [comment]
@mercibe Thank you for testing the proposed change. We understand your desire to see it be included 1.3.3, but, given that this code is fundamental to every Spring Boot application that's using an executable jar or war, we don't want to run the risk of introducing a regression in a maintenance release. We're going to include in in 1.4.0.M1.
